File: src/main/java/io/vproxy/vfx/util/imagewrapper/FX2BufferedImageCanvasBox.java
package io.vproxy.vfx.util.imagewrapper;

import javafx.embed.swing.SwingFXUtils;
import javafx.scene.image.WritableImage;

import java.awt.image.BufferedImage;

public class FX2BufferedImageCanvasBox implements CanvasBox {
    private final WritableImage img;
    private final BufferedImage bImg;
    private final BufferedImageCanvasBox box;
    private final double scale;

    public FX2BufferedImageCanvasBox(WritableImage img, double scale) {
        this.img = img;
        this.bImg = SwingFXUtils.fromFXImage(img, null);
        this.box = new BufferedImageCanvasBox(bImg, bImg.createGraphics());
        this.scale = scale;
    }

    @Override
    public void setPaint(int rgb) {
        box.setPaint(rgb);
    }

    @Override
    public void fillOval(int x, int y, int w, int h) {
        box.fillOval((int) (x / scale), (int) (y / scale), (int) (w / scale), (int) (h / scale));
    }

    @Override
    public void setFont(String name, boolean bold, int size) {
        box.setFont(name, bold, (int) (size / scale));
    }

    @Override
    public void drawString(String s, int x, int y) {
        box.drawString(s, (int) (x / scale), (int) (y / scale));
    }

    @Override
    public void flush() {
        box.flush();
        SwingFXUtils.toFXImage(bImg, img);
    }
}

File: src-ts/src/util/imagewrapper/FX2BufferedImageCanvasBox.ts
import type { CanvasBox } from './CanvasBox.js';
import { BufferedImageCanvasBox } from './BufferedImageCanvasBox.js';

export class FX2BufferedImageCanvasBox implements CanvasBox {
  private readonly box: BufferedImageCanvasBox;
  private readonly scale: number;

  constructor(writableCanvas: HTMLCanvasElement, scale: number) {
    const ctx = writableCanvas.getContext('2d');
    if (!ctx) {
      throw new Error('FX2BufferedImageCanvasBox: 2D context unavailable for canvas');
    }
    this.box = new BufferedImageCanvasBox(writableCanvas, ctx);
    this.scale = scale;
  }

  setPaint(rgb: number): void {
    this.box.setPaint(rgb);
  }

  fillOval(x: number, y: number, w: number, h: number): void {
    this.box.fillOval(
      Math.floor(x / this.scale),
      Math.floor(y / this.scale),
      Math.floor(w / this.scale),
      Math.floor(h / this.scale),
    );
  }

  setFont(name: string, bold: boolean, size: number): void {
    this.box.setFont(name, bold, Math.floor(size / this.scale));
  }

  drawString(s: string, x: number, y: number): void {
    this.box.drawString(s, Math.floor(x / this.scale), Math.floor(y / this.scale));
  }

  flush(): void {
    // No-op. Java copies pixels back to the WritableImage here; in the DOM
    // port we already drew directly onto the writable canvas.
    this.box.flush();
  }
}
